### ISSUE TITLE
Advanced Data Dictionary

### ISSUE THREAD

**Issue by jqnatividad** [dathere/datapusher-plus#18]

**Comment by wardi** [comment]
I like the idea of adding the jsonschema generated by qsv to the resource as a "schema" metadata field like the way ckanext-validation does. This way there's a clear method to download and reuse the schema, and it might even be interoperable with other extensions.

I don't like the data dictionary option because this data is all generated on upload so populating the data dictionary with these fields would be mix generated data together with user-supplied metadata. (would you make them read-only in the form? what about in the api?)

**Comment by jqnatividad** [comment]
I agree.The jsonschema produced by qsv is really robust. It enforces pretty much all the validation rules as per the jsonschema spec, and the generated schema can even have the enumerated constraints inferred by `qsv frequency` for a given threshold. It can even infer a regex pattern from the domain values of a column and add it to the jsonschema (https://github.com/jqnatividad/qsv/blob/master/src/cmd/schema.rs#L23).

Validating with `qsv validate` is also quite fast as its multi-threaded.  When we were testing it, we were validating a CSV against a non-trivial jsonschema spec at 300k records/second (https://github.com/jqnatividad/qsv/issues/164). 

And as you pointed out, the jsonschema can even be used by other extensions, as the jsonschema validation engine we use has [python bindings](https://github.com/Stranger6667/jsonschema-rs/tree/master/bindings/python).

As for the Data Dictionary UI, our client actually wants to be able to specify some things manually beyond just changing the description. For the alpha launch this summer, we're just going to stay with the current Data Dictionary form, and add the jsonschema file as a "system resource."

But yeah, ultimately, having some form of jsonschema editor would be nice. At the very least, to change/sync the description. Maybe, there can be a "power user" view they can switch to where they can manually edit the jsonschema, and just do a json lint check before saving the edits...

As for the Data Dictionary API, was thinking we can keep current interface for backwards compatibility and just document that there is a jsonschema file at a standard URL endpoint for each resource...

WDYT?

**Comment by wardi** [comment]
ckanext-validation stores the jsonschema as a resource metadata field so it can be edited as part of the form or via the API. I'm not sure what you mean by "system resource", is it stored in the db or on the filesystem or using an installed IUploader extension? what APIs are available for updating the contents?

**Comment by jqnatividad** [comment]
I'll have to look into ckanext-validation and understand how it works, and perhaps "borrow" some ideas. Do you have it running in production?

In my mind, a "system resource" is nothing more than a regular resource that's created by DP+ when pushing a "regular" resource with a special suffix (RESOURCEID_stats, RESOURCEID_freq, etc.), so it can be queried with the API just like any other resource.

**Comment by wardi** [comment]
So a resource that appears next to the others in the dataset? Or just some tables in the datastore db with corresponding table names?

We are using ckanext-validation in production on an internal version of our site.

**Comment by jqnatividad** [comment]
I was thinking it will just be a resource that appears next to the "regular resources". In that way, they're still addressable with the Datastore API... but there might be some advantages in storing them as regular PosgtgreSQL tables...

**Comment by wardi** [comment]
If they're regular resources I can see issues with users:
- wanting them displayed differently or hidden from the form/UI
- reordering them so it's not clear which system resource corresponds to which resource
- making changes to them that will get lost the next time the original resource changes

Or the extension itself might have problems creating system resources due to a customized resource metadata schema.

A specially named table in the datastore might be the way to go. Next add a custom endpoint that generates jsonschema based on data in a special table, and hook resource deletion so that these special tables are cleaned up. That should be most of the features you would want.

**Comment by jqnatividad** [comment]
Hi @wardi, in the 5/5/2022 dev team call, you mentioned you were working on a Table Designer.

Is that in a publicly available repo I can check?

**Comment by wardi** [comment]
just the branch I started a while ago where you can see some of the data dictionary overrides: https://github.com/ckan/ckan/compare/6118-table-designer#diff-9de4be8179eb39da879b2e19646f874128c85c2d673d277735c709405d706eac

**Comment by jqnatividad** [comment]
For the 0.4.0 release, DP+ sets the Data Dictionary Label field to the original column name as it was uploaded by the user.

This is because DP+ sanitizes the column names to be "db-safe" (i.e. valid postgres identifiers) with the qsv safenames command.

This pretty much guarantees a table will be created in the Datastore even if a CSV/Excel file has invalid column names.

By setting the Data Dictionary label to the original column name, they'll still see the original name with DataTables view as it uses the Data Dictionary label when its defined instead of the underlying postgresql column name.

**Comment by jqnatividad** [comment]
For the next release, DP+ will now have a setting SUMMARY_STATISTICS with a default value of false.

If set, a detailed summary statistics CSV file will be added as an additional resources with the "-stats" filename suffix.

**Comment by jqnatividad** [comment]
Just to update, DP+ has to option to add summary statistics to a resource since 0.5.0.

Here's the relevant section in the .env file

https://github.com/dathere/datapusher-plus/blob/48ca3eb522cb911683727403d95b8d6fa926f34f/datapusher/dot-env.template#L116-L128

**Comment by jqnatividad** [comment]
The next version will have the Data Dictionary stored as JSON Schema now that qsv describegpt supports JSON Schema as an output format.

https://github.com/dathere/qsv/pull/3871

Note that it's a valid JSON Schema that can be used for validation as we take advantage of JSON Schema's `additionalProperties`.
